File: frontend/src/types/certification.ts
export interface ModuleProgress {
  completed: boolean
  stars: number
  completed_at: string | null
  attempts: number
  xp_earned: number
  provisioned_docs?: string[]
  self_assessment?: Record<string, string>
}

export interface CertExercise {
  overview?: string
  documents: string[]
  instructions: string[]
  expected_fields: string[]
  expected_values: Record<string, string[]>
  star_criteria: Record<string, string>
}

export interface CertificationProgress {
  id: string
  user_id: string
  modules: Record<string, ModuleProgress>
  total_xp: number
  level: string
  certified: boolean
  certified_at: string | null
  streak_days: number
  last_activity_date: string | null
}

export interface ValidationCheck {
  name: string
  passed: boolean
  detail: string
}

export interface ValidationResult {
  passed: boolean
  stars: number
  checks: ValidationCheck[]
}

export interface CompletionResult {
  module_id: string
  stars: number
  xp_earned: number
  total_xp: number
  level: string
  level_up: boolean
  certified: boolean
  validation: ValidationResult
}

export interface KnowledgeCheckOption {
  text: string
  correct: boolean
  explanation: string
}

export interface KnowledgeCheckData {
  question: string
  options: KnowledgeCheckOption[]
}

export interface LessonSection {
  title: string
  content: string
  variant: 'concept' | 'walkthrough' | 'key-terms' | 'insight'
  knowledgeCheck?: KnowledgeCheckData
  diagram?: string
}

export interface ModuleDefinition {
  id: string
  number: number
  title: string
  subtitle: string
  description: string
  objectives: string[]
  tips: string[]
  lessons: LessonSection[]
  xp: number
  icon: string
  estimatedMinutes?: number
}

export interface TierDefinition {
  name: string
  theme: string
  narrative: string
  moduleIds: string[]
  celebration: string
}

File: backend/app/models/certification.py
import datetime
from typing import Optional

from beanie import Document


class CertificationProgress(Document):
    """Tracks a user's progress through the Vandal Workflow Architect certification."""

    user_id: str
    modules: dict = {}  # {module_id: {completed, stars, completed_at, attempts, xp_earned}}
    total_xp: int = 0
    level: str = "novice"
    certified: bool = False
    certified_at: Optional[datetime.datetime] = None
    streak_days: int = 0
    last_activity_date: Optional[str] = None  # YYYY-MM-DD for streak tracking
    created_at: datetime.datetime = datetime.datetime.now(tz=datetime.timezone.utc)
    updated_at: datetime.datetime = datetime.datetime.now(tz=datetime.timezone.utc)

    class Settings:
        name = "certification_progress"
